redb-derive/src/lib.rs

```diff
@@ -0,0 +1,156 @@
+pub fn derive_value(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+
+    match expand_derive_value(input) {
+        Ok(tokens) => tokens.into(),
+        Err(err) => err.to_compile_error().into(),
+    }
+}
+
+fn expand_derive_value(input: DeriveInput) -> syn::Result<proc_macro2::TokenStream> {
+    let struct_name = &input.ident;
+
+    let type_name = extract_type_name(&input.attrs)?;
+
+    let fields = match &input.data {
+        Data::Struct(data_struct) => match &data_struct.fields {
+            Fields::Named(fields) => &fields.named,
+            Fields::Unnamed(fields) => &fields.unnamed,
+            Fields::Unit => {
+                return Err(syn::Error::new_spanned(
+                    struct_name,
+                    "Unit structs are not supported",
+                ));
+            }
+        },
+        _ => {
+            return Err(syn::Error::new_spanned(
+                struct_name,
+                "Only structs are supported",
+            ));
+        }
+    };
+
+    if fields.is_empty() {
+        return Err(syn::Error::new_spanned(
+            struct_name,
+            "Empty structs are not supported",
+        ));
+    }
+
+    let field_types: Vec<_> = fields.iter().map(|f| &f.ty).collect();
+    let field_count = fields.len();
+
+    let tuple_type = if field_count == 1 {
+        let ty = &field_types[0];
+        quote! { (#ty,) }
+    } else {
+        quote! { (#(#field_types),*) }
+    };
+
+    let is_named_fields = fields.iter().next().unwrap().ident.is_some();
+
+    let (struct_to_tuple, tuple_to_struct) = if is_named_fields {
+        let field_names: Vec<_> = fields.iter().map(|f| f.ident.as_ref().unwrap()).collect();
+        let struct_to_tuple = if field_count == 1 {
+            let field = &field_names[0];
+            quote! { (&value.#field,) }
+        } else {
+            quote! { (#(&value.#field_names),*) }
+        };
+
+        let tuple_to_struct = {
+            let indices: Vec<Index> = (0..field_count).map(Index::from).collect();
+            quote! {
+                #struct_name {
+                    #(#field_names: tuple.#indices.clone()),*
+                }
+            }
+        };
+        (struct_to_tuple, tuple_to_struct)
+    } else {
+        let indices: Vec<Index> = (0..field_count).map(Index::from).collect();
+        let struct_to_tuple = if field_count == 1 {
+            quote! { (&value.0,) }
+        } else {
+            quote! { (#(&value.#indices),*) }
+        };
+
+        let tuple_to_struct = quote! {
+            #struct_name(#(tuple.#indices.clone()),*)
+        };
+        (struct_to_tuple, tuple_to_struct)
+    };
+
+    let type_name_with_fields = {
+        let field_type_names = field_types.iter().map(|ty| {
+            quote! { stringify!(#ty) }
+        });
+        quote! {
+            format!("{}({})", #type_name, [#(#field_type_names),*].join(","))
+        }
+    };
```

Comment:
When the fields have names, make the type name have this form: `"#type_name {#field_name: #field_type, #field2_name: #field2_type,...}"` so that it is more like the syntax of a Rust struct.

When the fields are unnamed then let's use the form you implemented: `#type_name (#field_type,...)`